### PR TITLE
fix: TextIteratorStreamer cannot work with pipeline

### LINKED ISSUE #23552: TextIteratorStreamer cannot be used with TextGenerationPipeline
### System Info

- `transformers` version: 4.29.2
- Platform: Linux-5.19.0-41-generic-x86_64-with-glibc2.35
- Python version: 3.8.13
- Huggingface_hub version: 0.14.1
- Safetensors version: not installed
- PyTorch version (GPU?): 2.0.0+cu117 (True)
- Tensorflow version (GPU?): 2.12.0 (True)
- Flax version (CPU?/GPU?/TPU?): 0.5.3 (cpu)
- Jax version: 0.4.10
- JaxLib version: 0.4.10
- Using GPU in script?: No
- Using distributed or parallel set-up in script?: No

### Who can help?

@Narsil @gante 

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

This issue occurs because the `TextIteratorStreamer` class contains a `Queue` field which cannot be pickled and the text generation pipeline runs a deepcopy .
https://github.com/huggingface/transformers/blob/3658488ff77ff8d45101293e749263acf437f4d5/src/transformers/pipelines/text_generation.py#L245

Code to reproduce issue:
```python
from transformers import AutoModelForCausalLM, AutoTokenizer, TextIteratorStreamer, pipeline

model = AutoModelForCausalLM.from_pretrained("gpt2")
tokenizer = AutoTokenizer.from_pretrained("gpt2")
streamer = TextIteratorStreamer(tokenizer)
pipe = pipeline(
    "text-generation", model=model, tokenizer=tokenizer, streamer=streamer
)
pipe("test")
```

Trace:
```python
Traceback (most recent call last):
  File "/home/raf/.pyenv/versions/3.8.13/lib/python3.8/threading.py", line 932, in _bootstrap_inner
    self.run()
  File "/home/raf/.pyenv/versions/3.8.13/lib/python3.8/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "/home/raf/.pyenv/versions/develop/lib/python3.8/site-packages/transformers/pipelines/text_generation.py", line 201, in __call__
    return super().__call__(text_inputs, **kwargs)
  File "/home/raf/.pyenv/versions/develop/lib/python3.8/site-packages/transformers/pipelines/base.py", line 1119, in __call__
    return self.run_single(inputs, preprocess_params, forward_params, postprocess_params)
  File "/home/raf/.pyenv/versions/develop/lib/python3.8/site-packages/transformers/pipelines/base.py", line 1126, in run_single
    model_outputs = self.forward(model_inputs, **forward_params)
  File "/home/raf/.pyenv/versions/develop/lib/python3.8/site-packages/transformers/pipelines/base.py", line 1025, in forward
    model_outputs = self._forward(model_inputs, **forward_params)
  File "/home/raf/.pyenv/versions/develop/lib/python3.8/site-packages/transformers/pipelines/text_generation.py", line 245, in _forward
    generate_kwargs = copy.deepcopy(generate_kwargs)
  File "/home/raf/.pyenv/versions/3.8.13/lib/python3.8/copy.py", line 146, in deepcopy
    y = copier(x, memo)
  File "/home/raf/.pyenv/versions/3.8.13/lib/python3.8/copy.py", line 230, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/home/raf/.pyenv/versions/3.8.13/lib/python3.8/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/home/raf/.pyenv/versions/3.8.13/lib/python3.8/copy.py", line 270, in _reconstruct
    state = deepcopy(state, memo)
  File "/home/raf/.pyenv/versions/3.8.13/lib/python3.8/copy.py", line 146, in deepcopy
    y = copier(x, memo)
  File "/home/raf/.pyenv/versions/3.8.13/lib/python3.8/copy.py", line 230, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/home/raf/.pyenv/versions/3.8.13/lib/python3.8/copy.py", line 172, in deepcopy
    y = _reconstruct(x, memo, *rv)
  File "/home/raf/.pyenv/versions/3.8.13/lib/python3.8/copy.py", line 270, in _reconstruct
    state = deepcopy(state, memo)
  File "/home/raf/.pyenv/versions/3.8.13/lib/python3.8/copy.py", line 146, in deepcopy
    y = copier(x, memo)
  File "/home/raf/.pyenv/versions/3.8.13/lib/python3.8/copy.py", line 230, in _deepcopy_dict
    y[deepcopy(key, memo)] = deepcopy(value, memo)
  File "/home/raf/.pyenv/versions/3.8.13/lib/python3.8/copy.py", line 161, in deepcopy
    rv = reductor(4)
TypeError: cannot pickle '_thread.lock' object

Process finished with exit code 137 (interrupted by signal 9: SIGKILL)
```

### Expected behavior

Pipeline should run normally

### PR DESCRIPTION
**Pull request by yuanwu2017**
Deepcopying the TextIteratorStreamer object causes the exception.

# What does this PR do?
Fixes #23552 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by yao-matrix** [comment]
@sgugger , pls help review, thx.

**Comment by yao-matrix** [comment]
@sywangyi

**Comment by yuanwu2017** [comment]
@gante Please help to review.

**Comment by opcode81** [comment]
I ran into the same issue and also had another issue about a week ago (which I don't remember now) that was also related to the deepcopy operation. If the sole purpose of the deep copy is to allow calls to `pop` without modifying the dictionary, then shouldn't a shallow copy do?

**Comment by yuanwu2017** [comment]
@gante  We used the transformers pipeline with TextIteratorStreamer for chatbot, but it raised the  deepcopy exception. Can you help to check the issue and review the PR?
Thanks a lot.

**Comment by sgugger** [comment]
@gante is on vacation, please be patient until he comes back :-)
Also cc @Narsil

**Comment by yuanwu2017** [comment]
> @gante is on vacation, please be patient until he comes back :-) Also cc @Narsil

Got it. Thanks.

File: src/transformers/pipelines/text_generation.py [hunk 1]
```diff
@@ -242,7 +242,12 @@ def _forward(self, model_inputs, **generate_kwargs):
 
         # If there is a prefix, we may need to adjust the generation length. Do so without permanently modifying
         # generate_kwargs, as some of the parameterization may come from the initialization of the pipeline.
+        streamer = None
+        if "streamer" in generate_kwargs.keys():
+            streamer = generate_kwargs.pop("streamer")
         generate_kwargs = copy.deepcopy(generate_kwargs)
+        if streamer:
+            generate_kwargs["streamer"] = streamer
```
Comment (gante) [inline src/transformers/pipelines/text_generation.py:250]:
```suggestion
        generate_kwargs = copy.copy(generate_kwargs)
```

I think we can make it a shallow copy (see [this comment](https://github.com/huggingface/transformers/issues/23552#issuecomment-1584205325))

**Review by gante — COMMENTED** [review]
Apologies for the long delay 🤗 Have a look at the suggestion I added, I believe it solves the problem while keeping it short!

**Comment by yuanwu2017** [comment]
Do I need to fix these test failures which are not related with this patch? @gante

File: src/transformers/pipelines/text_generation.py [hunk 2]
```diff
@@ -242,7 +242,7 @@ def _forward(self, model_inputs, **generate_kwargs):
 
         # If there is a prefix, we may need to adjust the generation length. Do so without permanently modifying
         # generate_kwargs, as some of the parameterization may come from the initialization of the pipeline.
-        generate_kwargs = copy.deepcopy(generate_kwargs)
+        generate_kwargs = copy.copy(generate_kwargs)
```
Comment (gante) [inline src/transformers/pipelines/text_generation.py:245]:
```suggestion
```

Even shorter version -- we don't need a copy at all, see the conversation [here](https://github.com/huggingface/transformers/issues/23785)

**Review by gante — APPROVED** [review]

**Comment by gante** [comment]
@yuanwu2017 to get rid of the CI errors, please rebase the PR. It is a hard requirement -- LMK if you need instructions to do so :)

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing!

**Pull request merged**